Add Switcher Runner S11 support

YogevBokobza opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/34269

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

YogevBokobza committed:
switcher start s11 integration

YogevBokobza committed:
switcher linting

YogevBokobza commented:
@thecode 
Regarding this in-progress PR I have a couple of questions that maybe you can help:
1. In previous PRs to switcher_kis there was a change of config_flow to this: `config_entry_flow.register_discovery_flow(DOMAIN, "Switcher", async_has_devices)` I need to revert that to a full setup file right? because right now we going to need auth having CONF_TOKEN and CONF_USERNAME
2. Please take a look at [this](https://github.com/home-assistant/core/pull/123578/files#diff-dde25b23ebea17f28ff7a697f8d0045b8508075777c403309b27583d3c36e712R59-R69), I am trying to use the API call from the `aioswithcer` library and I am getting that fail. How to fix that?
3. Regarding [this](https://github.com/home-assistant/core/pull/123578/files#diff-3e4eacfcf78f613ccf0f87e8409a1e040758f7e7ec9bcb802becd3de9458b96cR68), I don't like passing token as that.. I think after (1) I should get the token from the config_entry right?
4. Regarding [this logic](https://github.com/home-assistant/core/pull/123578/files#diff-3e4eacfcf78f613ccf0f87e8409a1e040758f7e7ec9bcb802becd3de9458b96cR58-R66), So in the discovery we found out a device that requires a token. Now it should prompt a form for the user to enter a username(which is email) and token(after he manually goes and fetches it from the Switcher URL), validate it and store the token so when creating each coordinator it will have the token in it.
You mentioned [reauthentication](https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#reauthentication), the documentation is really lacking on this and I could not find an appropriate example in other integration.. Can you assist with that?

thecode commented:
> 1. In previous PRs to switcher_kis there was a change of config_flow to this: `config_entry_flow.register_discovery_flow(DOMAIN, "Switcher", async_has_devices)` I need to revert that to a full setup file right? because right now we going to need auth having CONF_TOKEN and CONF_USERNAME

You need to implement a full config flow

> 2. Please take a look at [this](https://github.com/home-assistant/core/pull/123578/files#diff-dde25b23ebea17f28ff7a697f8d0045b8508075777c403309b27583d3c36e712R59-R69), I am trying to use the API call from the `aioswithcer` library and I am getting that fail. How to fix that?

Unfortunately this was implemented as a blocking call in `aioswitcher` so you either fix it there (best option would be to use `aiohttp`) or run it in executor. See [Blocking operations with asyncio
](https://developers.home-assistant.io/docs/asyncio_blocking_operations/)

> 3. Regarding [this](https://github.com/home-assistant/core/pull/123578/files#diff-3e4eacfcf78f613ccf0f87e8409a1e040758f7e7ec9bcb802becd3de9458b96cR68), I don't like passing token as that.. I think after (1) I should get the token from the config_entry right?

You can access the token from `entry.data` (`entry.data.get(CONF_TOKEN)`), no need to pass it to the coordinator

> 4. Regarding [this logic](https://github.com/home-assistant/core/pull/123578/files#diff-3e4eacfcf78f613ccf0f87e8409a1e040758f7e7ec9bcb802becd3de9458b96cR58-R66), So in the discovery we found out a device that requires a token. Now it should prompt a form for the user to enter a username(which is email) and token(after he manually goes and fetches it from the Switcher URL), validate it and store the token so when creating each coordinator it will have the token in it.
>    You mentioned [reauthentication](https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#reauthentication), the documentation is really lacking on this and I could not find an appropriate example in other integration.. Can you assist with that?

Here is an example for starting a reauth flow:
https://github.com/home-assistant/core/blob/a040f1a9d15621c8ae6f13fddea35e1ad6ede95f/homeassistant/components/shelly/coordinator.py#L717

You can start the reauth flow when you detect a new device here:
https://github.com/home-assistant/core/blob/a040f1a9d15621c8ae6f13fddea35e1ad6ede95f/homeassistant/components/switcher_kis/__init__.py#L47
and skip adding the device or generating coordinator for it. After the user will finish the flow the integration will be reloaded and devices will be added again.

YogevBokobza committed:
switcher starting reauth logic

YogevBokobza commented:
@thecode 
Made some progress
(1) Made full config flow (basically revert what it was before the last change of minimize that)
(2) Added CONF_USERNAME and CONF_TOKEN
(3) Added reauth step.. it's kinda buggy right now. Please take a look to see what is going on..
(4) I am skipping creating a coordinator if a token is needed but not provided.
(5) Regarding the validate token function I have no idea how to implement it at the moment..

Please revisit the changes and let me know what you think and how to improve.

YogevBokobza committed:
switcher fix linting

YogevBokobza committed:
switcher fix linting

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,97 @@
+        if self._async_current_entries(True):
+            return self.async_abort(reason="single_instance_allowed")
+
+        self.hass.data.setdefault(DOMAIN, {})
+        if DATA_DISCOVERY not in self.hass.data[DOMAIN]:
+            self.hass.data[DOMAIN][DATA_DISCOVERY] = self.hass.async_create_task(
+                async_discover_devices()
+            )
```
Please don't use previous config flow implementation in Switcher as an example for config flow. It is not allowed (and not needed) to store data in `hass.data` during config flow.

YogevBokobza replied:
So can you show me a good example?

YogevBokobza replied:
I revert that part back.. I dont know how to progress from here..

thecode replied:
The `async_discover_devices()` was the right direction, just don't store the discovered devices inside `self.hass.data`, you can store them in a temp variable (`discovered_devices`) if they are only used inside this method, if you need them in another method you can store them as a class variable (`self._discovered devices`)

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,97 @@
+
+        self.hass.data.setdefault(DOMAIN, {})
+        if DATA_DISCOVERY not in self.hass.data[DOMAIN]:
+            self.hass.data[DOMAIN][DATA_DISCOVERY] = self.hass.async_create_task(
```
This is already async, why does it run in a task?

YogevBokobza replied:
I just copied it over from the previous iteration of this file..

thecode replied:
You copied parts which are not related and mixed them, can  you remove the `async_create_task`?

YogevBokobza replied:
I revert that part back.. I dont know how to progress from here..

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,97 @@
+            data={
+                CONF_USERNAME: self.username,
+                CONF_TOKEN: self.token,
+            },
```
Please first check if one of the discovered devices needs a token before asking the user to add a token

YogevBokobza replied:
I do that here:
https://github.com/home-assistant/core/pull/123578/files#diff-3e4eacfcf78f613ccf0f87e8409a1e040758f7e7ec9bcb802becd3de9458b96cR57-R61
I am calling reauth to get username and token only if device that needs a token found

thecode replied:
This needs to happen in 2 places:
1. Same place you already added, this will handle devices that are detected during runtime that needs a token.
2. When user add the integration, if a device that needs a token is detect you need to switch to asking for credentials before continuing to create an entry.

Exampe:
https://github.com/home-assistant/core/blob/21987a67e7c4d3c526ddfbd4866c64f93c2569f0/homeassistant/components/shelly/config_flow.py#L171-L172

YogevBokobza replied:
> This needs to happen in 2 places:
> 
> 1. Same place you already added, this will handle devices that are detected during runtime that needs a token.
> 2. When user add the integration, if a device that needs a token is detect you need to switch to asking for credentials before continuing to create an entry.
> 
> Exampe:
> 
> https://github.com/home-assistant/core/blob/21987a67e7c4d3c526ddfbd4866c64f93c2569f0/homeassistant/components/shelly/config_flow.py#L171-L172

Maybe I am missing something but when I tested this part I got the "reauth" when I initially set the integration for the first time and when I had this already set up and I pressed reload on the integration (And when I didn't have the device there and restart HA)..

Plus this is the only part where the bridge is detecting the devices so I can determine if this is a token-based device no?

YogevBokobza replied:
I put the missing step and did a little refactoring. Please revisit.

thecode commented on homeassistant/components/switcher_kis/const.py:
```diff
@@ -1,10 +1,15 @@
+CONF_TOKEN = "token"
+
+COVER1_ID = "runner"
+COVER2_ID = "runner2"
```
Is it going to be used anywhere in this PR? if it is for future device add it when you add the support for that device

YogevBokobza replied:
I need the COVER1_ID but removed the COVER2_ID

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -117,16 +134,20 @@ async def _async_call_api(self, api: str, *args: Any) -> None:
     async def async_close_cover(self, **kwargs: Any) -> None:
         """Close cover."""
-        await self._async_call_api(API_SET_POSITON, 0)
+        index = get_circuit_number(self._cover_id)
+        await self._async_call_api(API_SET_POSITON, 0, index)
```
The implementation of the `get_circuit_number` here does not make sense, however since it is not used in this PR (it will simply return `0`) please remove it and remove the `index` parameter, you can add it in a future PR when you add support for dual cover.

YogevBokobza replied:
Removed `get_circuit_number ` and `index`

thecode commented on homeassistant/components/switcher_kis/strings.json:
```diff
@@ -3,6 +3,13 @@
         "description": "[%key:common::config_flow::description::confirm_setup%]"
+      },
+      "reauth_confirm": {
+        "description": "Found a Switcher device that requires a token\nEnter your username(Email Address) and token here\nFor security reasons, some devices require a token to operate,\nthe token can be acquired from this link: https://switcher.co.il/GetKey/\nExample of Token: zvVvd7JxtN7CgvkD1Psujw==",
```
Please refer to a new section in the docs instead of the URL and example

YogevBokobza replied:
@thecode 
I don't understand what doc is you talking about?

YogevBokobza replied:
@thecode 
You mean I need to make a PR [here ](https://github.com/home-assistant/home-assistant.io/blob/9374fd5bc2fbcab87431bf5327135257133c0f99/source/_integrations/switcher_kis.markdown?plain=1)?
How is it working there? Should It be merged there and here at the same time?

thecode replied:
Yes, don't worry, they will be merged together

YogevBokobza replied:
Fix the strings.. please see now

YogevBokobza committed:
switcher remove get_circuit_number

YogevBokobza committed:
switcher adding support for validate token

YogevBokobza committed:
switcher fix initial auth for new devices and fix strings

YogevBokobza committed:
switcher fix linting

YogevBokobza committed:
switcher fix utils

YogevBokobza committed:
Revert "switcher fix utils"

This reverts commit b162a943b94fb0a581140feb21fe871df578c16a.

YogevBokobza committed:
switcher revert and test

YogevBokobza commented:
@thecode 
I tried to use this edited [function](https://github.com/home-assistant/core/pull/123578/files#diff-dde25b23ebea17f28ff7a697f8d0045b8508075777c403309b27583d3c36e712R48-R55)
But when I used that in config_flow it resulted in some error from the home assistant core. Do you know why?

YogevBokobza commented:
Also, I have this warning when pressing on the confirm for username and token:

> 2024-08-13 21:37:08.053 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to load_default_certs with args (<ssl.SSLContext object at 0x7f0583ca1350>, <Purpose.SERVER_AUTH: _ASN1Object(nid=129, shortname='serverAuth', longname='TLS Web Server Authentication', oid='1.3.6.1.5.5.7.3.1')>) inside the event loop by integration 'switcher_kis' at homeassistant/components/switcher_kis/config_flow.py, line 71: token_is_valid = await validate_token( (offender: /usr/local/lib/python3.12/ssl.py, line 713: context.load_default_certs(purpose)), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+switcher_kis%22
> For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#load_default_certs
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.12/runpy.py", line 198, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/usr/local/lib/python3.12/runpy.py", line 88, in _run_code
>     exec(code, run_globals)
>   File "/home/vscode/.vscode-server/extensions/ms-python.debugpy-2024.10.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
>     cli.main()
>   File "/home/vscode/.vscode-server/extensions/ms-python.debugpy-2024.10.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
>     run()
>   File "/home/vscode/.vscode-server/extensions/ms-python.debugpy-2024.10.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 317, in run_module
>     run_module_as_main(options.target, alter_argv=True)
>   File "/home/vscode/.vscode-server/extensions/ms-python.debugpy-2024.10.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 238, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/home/vscode/.vscode-server/extensions/ms-python.debugpy-2024.10.0-linux-x64/bundled/libs/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
>     exec(code, run_globals)
>   File "/workspaces/core/homeassistant/__main__.py", line 223, in <module>
>     sys.exit(main())
>   File "/workspaces/core/homeassistant/__main__.py", line 209, in main
>     exit_code = runner.run(runtime_conf)
>   File "/workspaces/core/homeassistant/runner.py", line 190, in run
>     return loop.run_until_complete(setup_and_run_hass(runtime_config))
>   File "/usr/local/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete
>     self.run_forever()
>   File "/usr/local/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
>     self._run_once()
>   File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1979, in _run_once
>     handle._run()
>   File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
>     self._context.run(self._callback, *self._args)
>   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_protocol.py", line 459, in _handle_request
>     resp = await request_handler(request)
>   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_app.py", line 537, in _handle
>     resp = await handler(request)
>   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_middlewares.py", line 114, in impl
>     return await handler(request)
>   File "/workspaces/core/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware
>     return await handler(request)
>   File "/workspaces/core/homeassistant/components/http/forwarded.py", line 83, in forwarded_middleware
>     return await handler(request)
>   File "/workspaces/core/homeassistant/components/http/request_context.py", line 26, in request_context_middleware
>     return await handler(request)
>   File "/workspaces/core/homeassistant/components/http/ban.py", line 85, in ban_middleware
>     return await handler(request)
>   File "/workspaces/core/homeassistant/components/http/auth.py", line 242, in auth_middleware
>     return await handler(request)
>   File "/workspaces/core/homeassistant/components/http/headers.py", line 32, in headers_middleware
>     response = await handler(request)
>   File "/workspaces/core/homeassistant/helpers/http.py", line 73, in handle
>     result = await handler(request, **request.match_info)
>   File "/workspaces/core/homeassistant/components/http/decorators.py", line 81, in with_admin
>     return await func(self, request, *args, **kwargs)
>   File "/workspaces/core/homeassistant/components/config/config_entries.py", line 222, in post
>     return await super().post(request, flow_id)
>   File "/workspaces/core/homeassistant/components/http/data_validator.py", line 74, in wrapper
>     return await method(view, request, data, *args, **kwargs)
>   File "/workspaces/core/homeassistant/helpers/data_entry_flow.py", line 122, in post
>     result = await self._flow_mgr.async_configure(flow_id, data)
>   File "/workspaces/core/homeassistant/data_entry_flow.py", line 366, in async_configure
>     result = await self._async_configure(flow_id, user_input)
>   File "/workspaces/core/homeassistant/data_entry_flow.py", line 413, in _async_configure
>     result = await self._async_handle_step(
>   File "/workspaces/core/homeassistant/data_entry_flow.py", line 516, in _async_handle_step
>     result: _FlowResultT = await getattr(flow, method)(user_input)
>   File "/workspaces/core/homeassistant/components/switcher_kis/config_flow.py", line 71, in async_step_credentials
>     token_is_valid = await validate_token(
> 
> 2024-08-13 21:37:08.116 WARNING (MainThread) [asyncio] Executing <Task pending name='Task-108' coro=<RequestHandler._handle_request() running at /home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_protocol.py:459> wait_for=<Future pending cb=[shield.<locals>._outer_done_callback() at /usr/local/lib/python3.12/asyncio/tasks.py:922, Task.task_wakeup()] created at /usr/local/lib/python3.12/asyncio/base_events.py:449> cb=[Task.task_wakeup()] created at /home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_protocol.py:536> took 0.150 seconds
> 

Can you see why?

YogevBokobza committed:
switcher fix validate logic and strings

YogevBokobza committed:
switcher add tests to improve coverage

YogevBokobza committed:
switcher adding tests

YogevBokobza committed:
switcher adding test

YogevBokobza commented:
I am quite stuck here.. I don't know if I am doing things right and having trouble with the tests and the coverage..

thecode commented on homeassistant/components/switcher_kis/const.py:
```diff
@@ -1,10 +1,14 @@
 
 DOMAIN = "switcher_kis"
-
+DATA_DISCOVERY = "discovery"
```
```suggestion
```
Please remove this

YogevBokobza replied:
Removed.

YogevBokobza committed:
switcher revert back things

YogevBokobza commented:
@thecode I will be happy if you can assist here and guide me on how to progress..

thecode commented on homeassistant/components/switcher_kis/utils.py:
```diff
@@ -42,3 +43,13 @@ def on_device_data_callback(device: SwitcherBase) -> None:
     return SwitcherBreezeRemoteManager()
+
+
+async def validate_input(username: str, token: str) -> bool:
+    """Validate token by specifying username and token."""
+    token_is_valid = await validate_token(username, token)
+    if token_is_valid:
+        _LOGGER.info("Token is valid")
+        return True
+    _LOGGER.info("Token is invalid")
+    return False
```
Why do we need a new method here which just return the same data from `validate_token`? please use it directly where needed. Also avoid using logger in `info` level. 
Here there is no need for logging the result as it will be visible to the user when he enter the token.

thecode commented on tests/components/switcher_kis/__init__.py:
```diff
@@ -1,14 +1,27 @@
-    entry = MockConfigEntry(domain=DOMAIN, data={}, unique_id=DOMAIN)
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={
+            CONF_USERNAME: username,
+            CONF_TOKEN: token,
+        },
```
Add the `username` and `token` to the data only if they are not `none`. This is not a valid mocking as existing entries will not have these fields at all and new entries will have them with a value.
Example:
https://github.com/home-assistant/core/blob/41b129e990d9c2993dab3e34e9df37faa1059856/tests/components/shelly/__init__.py#L50-L51

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -44,6 +44,11 @@ def async_add_cover(coordinator: SwitcherDataUpdateCoordinator) -> None:
+            coordinator.data.device_type.category
+            == DeviceCategory.SINGLE_SHUTTER_DUAL_LIGHT
+        ):
+            async_add_entities([SwitcherCoverEntity(coordinator, COVER1_ID)])
```
```suggestion
            async_add_entities([SwitcherCoverEntity(coordinator, 0)])
```
Instead of using a string, pass here the cover index which is more reflecting the physical device, than you can later use this directly in the API calls. Future devices will pass the correct index also (1, 2)

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -65,11 +70,21 @@ class SwitcherCoverEntity(
+    def __init__(
+        self,
+        coordinator: SwitcherDataUpdateCoordinator,
+        cover_id: str | None = None,
```
```suggestion
        cover_id: int| None = None,
```

YogevBokobza committed:
switcher fix based on requested changes

thecode commented on homeassistant/components/switcher_kis/__init__.py:
```diff
@@ -45,14 +45,21 @@ def on_device_data_callback(device: SwitcherBase) -> None:
+            device.token_needed,
         )
 
+        if device.token_needed:
+            _LOGGER.info("Found a Switcher device that needs a token")
+            if not entry.data.get(CONF_TOKEN):
+                entry.async_start_reauth(hass)
+                return
```
```suggestion
        if device.token_needed and not entry.data.get(CONF_TOKEN):
            entry.async_start_reauth(hass)
            return
```

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,125 @@
+    entry: ConfigEntry | None = None
+    username: str | None = None
+    token: str | None = None
+    discovered_devices: dict[str, SwitcherBase] | None = None
```
```suggestion
    discovered_devices: dict[str, SwitcherBase] = {}
```

YogevBokobza committed:
switcher tests fixes

YogevBokobza committed:
switcher fix based on requested changes

YogevBokobza commented:
@thecode  Done all the recently requested changes.. See if you can find more stuff to change.
Thanks.

YogevBokobza commented:
@thecode
I am trying to solve the coverage part.
I am getting this:
Name                                                   Stmts   Miss  Cover   Missing
------------------------------------------------------------------------------------
homeassistant/components/switcher_kis/__init__.py         40      0   100%
homeassistant/components/switcher_kis/button.py           53      0   100%
homeassistant/components/switcher_kis/climate.py         115      2    98%   206, 223
homeassistant/components/switcher_kis/config_flow.py      57     18    68%   36, 51-52, 59-78, 97-104
homeassistant/components/switcher_kis/const.py            10      0   100%
homeassistant/components/switcher_kis/coordinator.py      32      0   100%
homeassistant/components/switcher_kis/cover.py            69      0   100%
homeassistant/components/switcher_kis/diagnostics.py      10      0   100%
homeassistant/components/switcher_kis/sensor.py           36      0   100%
homeassistant/components/switcher_kis/switch.py           86      0   100%
homeassistant/components/switcher_kis/utils.py            26      0   100%
------------------------------------------------------------------------------------
TOTAL                                                    534     20    96%

The climate one was like this before..

I am trying to solve those:
homeassistant/components/switcher_kis/config_flow.py      57     18    68%   36, 51-52, 59-78, 97-104

Line 36: I don't know why the current `test_single_instance` not working
For lines 51-52: How can I test `discovered_devices` ? You said it should not be part of `hass.data` but then how can I test it?

thecode commented:
> Line 36: I don't know why the current `test_single_instance` not working For lines 51-52: How can I test `discovered_devices` ? 

It is not working because this integration is already a single config entry integration: https://github.com/home-assistant/core/blob/ada6b7875c2a7f4a54ba9dd93c70a93047d71874/homeassistant/components/switcher_kis/manifest.json#L11

https://developers.home-assistant.io/docs/creating_integration_manifest#single-config-entry-only

YogevBokobza commented:
> > Line 36: I don't know why the current `test_single_instance` not working For lines 51-52: How can I test `discovered_devices` ?
> 
> It is not working because this integration is already a single config entry integration:
> 
> https://github.com/home-assistant/core/blob/ada6b7875c2a7f4a54ba9dd93c70a93047d71874/homeassistant/components/switcher_kis/manifest.json#L11
> 
> https://developers.home-assistant.io/docs/creating_integration_manifest#single-config-entry-only

So, what do you think I should do? remove the `"single_config_entry": true` or just remove that check part from the config_flow?

thecode commented:
> So, what do you think I should do? remove the `"single_config_entry": true` or just remove that check part from the config_flow?

Was there any reason for adding this part to config flow?

YogevBokobza commented:
> > So, what do you think I should do? remove the `"single_config_entry": true` or just remove that check part from the config_flow?
> 
> Was there any reason for adding this part to config flow?

I guess I needed that for this confirm part https://github.com/home-assistant/core/pull/123578/files#diff-0abb13b727d4e23d4e4858b16667de7a42cba44c2d303db0deb59b4bdaee2902R42-R53 no?

YogevBokobza commented:
@thecode
I need your help with this..
I am trying this code:
```
async def test_reauth_successful(hass: HomeAssistant) -> None:
    """Test starting a reauthentication flow."""

    mock_config = MockConfigEntry(
        domain=DOMAIN, data={CONF_USERNAME: DUMMY_USERNAME, CONF_TOKEN: DUMMY_TOKEN}
    )
    mock_config.add_to_hass(hass)

    # Create a mock response object
    mock_response = AsyncMock()
    mock_response.status = 200
    mock_response.json = AsyncMock(return_value={"result": "true"})

    # Create an async mock for aiohttp.ClientSession
    mock_session = AsyncMock()
    mock_session.post = AsyncMock(return_value=mock_response)

    # Configure the async context manager methods for mock_session
    mock_session.__aenter__.return_value = mock_session
    mock_session.__aexit__.return_value = AsyncMock()

    # Patch aiohttp.ClientSession to use our mock
    with (
        patch("aiohttp.ClientSession", new=AsyncMock(return_value=mock_session)),
        patch("homeassistant.components.switcher_kis.async_setup_entry"),
        patch("aioswitcher.device.tools.validate_token", return_value=True),
    ):
        result = await hass.config_entries.flow.async_init(
            DOMAIN,
            context={
                "source": config_entries.SOURCE_REAUTH,
                "entry_id": mock_config.entry_id,
            },
            data=mock_config.data,
        )

        assert result["type"] == FlowResultType.FORM
        assert result["step_id"] == "reauth_confirm"

        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            user_input={
                CONF_USERNAME: DUMMY_USERNAME,
                CONF_TOKEN: DUMMY_TOKEN,
            },
        )
        await hass.async_block_till_done()

        assert result["type"] == FlowResultType.ABORT
        assert result["reason"] == "reauth_successful"
```

I am keep getting this error:
```

tests/components/switcher_kis/test_config_flow.py::test_reauth_successful
  /home/vscode/.local/ha-venv/lib/python3.12/site-packages/aioswitcher/device/tools.py:203: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
  Coroutine created at (most recent call last)
    File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1979, in _run_once
      handle._run()
    File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
      self._context.run(self._callback, *self._args)
    File "/workspaces/core/tests/components/switcher_kis/test_config_flow.py", line 185, in test_reauth_successful
      result = await hass.config_entries.flow.async_configure(
    File "/workspaces/core/homeassistant/data_entry_flow.py", line 366, in async_configure
      result = await self._async_configure(flow_id, user_input)
    File "/workspaces/core/homeassistant/data_entry_flow.py", line 413, in _async_configure
      result = await self._async_handle_step(
    File "/workspaces/core/homeassistant/data_entry_flow.py", line 516, in _async_handle_step
      result: _FlowResultT = await getattr(flow, method)(user_input)
    File "/workspaces/core/homeassistant/components/switcher_kis/config_flow.py", line 98, in async_step_reauth_confirm
      token_is_valid = await validate_token(
    File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aioswitcher/device/tools.py", line 203, in validate_token
      async with aiohttp.ClientSession() as session:
    File "/usr/local/lib/python3.12/unittest/mock.py", line 1134, in __call__
      return self._mock_call(*args, **kwargs)
    File "/usr/local/lib/python3.12/unittest/mock.py", line 1138, in _mock_call
      return self._execute_mock_call(*args, **kwargs)
    async with aiohttp.ClientSession() as session:
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
  
  ```
  
  And this error:
  ```
  >       async with aiohttp.ClientSession() as session:
E       TypeError: 'coroutine' object does not support the asynchronous context manager protocol

/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aioswitcher/device/tools.py:203: TypeError
  ```
  
  Can you help me find out what is wrong?
  Thanks.

YogevBokobza committed:
switcher remove single_instance_allowed code and added tests

YogevBokobza committed:
Update config_flow.py

YogevBokobza commented:
@thecode 
I removed that [part ](https://github.com/home-assistant/core/pull/123578/commits/1b5da7049894a5d9ec8b9a9405112130c9091291#diff-0abb13b727d4e23d4e4858b16667de7a42cba44c2d303db0deb59b4bdaee2902R34-R35)

I also added tests..
My main issue is mocking `validate_token` I am still getting this [issue](https://github.com/home-assistant/core/pull/123578#issuecomment-2323530741).
Please help me figure this out..

Thanks.

YogevBokobza committed:
switcher fix comment

YogevBokobza committed:
switcher fix tests

YogevBokobza committed:
switcher lint

thecode commented on homeassistant/components/switcher_kis/__init__.py:
```diff
@@ -12,7 +12,7 @@
 from homeassistant.helpers import device_registry as dr
 
-from .const import DOMAIN
+from .const import CONF_TOKEN, DOMAIN
```
This token exists as core constant:
`from homeassistant.const import CONF_TOKEN`

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/__init__.py:
```diff
@@ -45,14 +45,19 @@ def on_device_data_callback(device: SwitcherBase) -> None:
+            device.token_needed,
         )
 
+        if device.token_needed and not entry.data.get(CONF_TOKEN):
```
Since this can happen asynchronously during startup please store the constant locally (at line ~30):
`token = entry.data.get(CONF_TOKEN)`

```suggestion
        if device.token_needed and not token:
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,119 @@
-config_entry_flow.register_discovery_flow(DOMAIN, "Switcher", async_has_devices)
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, ConfigFlowResult
+
+from .const import CONF_TOKEN, CONF_USERNAME, DOMAIN
```
Use core constants:
`from homeassistant.const import CONF_USERNAME, CONF_TOKEN`

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/const.py:
```diff
@@ -5,6 +5,8 @@
 DISCOVERY_TIME_SEC = 12
 
 SIGNAL_DEVICE_ADD = "switcher_device_add"
+CONF_USERNAME = "username"
+CONF_TOKEN = "token"
```
```suggestion
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/coordinator.py:
```diff
@@ -12,7 +12,7 @@
 from homeassistant.helpers.dispatcher import async_dispatcher_send
 
-from .const import DOMAIN, MAX_UPDATE_INTERVAL_SEC, SIGNAL_DEVICE_ADD
+from .const import CONF_TOKEN, DOMAIN, MAX_UPDATE_INTERVAL_SEC, SIGNAL_DEVICE_ADD
```
`from homeassistant.const import CONF_TOKEN`

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,119 @@
-config_entry_flow.register_discovery_flow(DOMAIN, "Switcher", async_has_devices)
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, ConfigFlowResult
+
+from .const import CONF_TOKEN, CONF_USERNAME, DOMAIN
```
From core constants

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,119 @@
+        else:
+            user_input = {}
+
+        schema = {
+            vol.Required(CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")): str,
+            vol.Required(CONF_TOKEN, default=user_input.get(CONF_TOKEN, "")): str,
+        }
```
Please move the schema as a constant outside of the class (since you use it in more than one place), example:
https://github.com/home-assistant/core/blob/50a46933f62c2495cdc5e781ee13aa872c98036b/homeassistant/components/shelly/config_flow.py#L60-L65

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,119 @@
+                )
+            errors["base"] = "invalid_auth"
+
+        schema = {
+            vol.Required(CONF_USERNAME): str,
+            vol.Required(CONF_TOKEN): str,
+        }
```
https://github.com/home-assistant/core/blob/50a46933f62c2495cdc5e781ee13aa872c98036b/homeassistant/components/shelly/config_flow.py#L60-L65

YogevBokobza replied:
It is done.

thecode reviewed (changes requested):
Overall looks good, I didn't review the tests yet and only reviewed part of the config flow, please fix the changes and I will review config flow again. Thanks

YogevBokobza committed:
switcehr fix based on requested changes

YogevBokobza committed:
switche fix lint

YogevBokobza committed:
switcher small rename fix

YogevBokobza commented:
@thecode 
Do you think it makes more sense to rename those tests function "test_cover2*" to something else?
like "test_cover_token" or something?
I think it will cause a confusion when I add Switcher Runner S12 device which it has 2 covers

thecode commented on homeassistant/components/switcher_kis/strings.json:
```diff
@@ -3,11 +3,29 @@
         "description": "[%key:common::config_flow::description::confirm_setup%]"
+      },
+      "credentials": {
+        "description": "Found a Switcher device that requires a token\nEnter your username(Email Address) and token here\nFor more information see here:\nhttps://www.home-assistant.io/integrations/switcher_kis/#prerequisites",
```
```suggestion
        "description": "Found a Switcher device that requires a token\nEnter your username and token\nFor more information see https://www.home-assistant.io/integrations/switcher_kis/#prerequisites",
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/strings.json:
```diff
@@ -3,11 +3,29 @@
+        }
+      },
+      "reauth_confirm": {
+        "description": "Found a Switcher device that requires a token\nEnter your username(Email Address) and token here\nFor more information see here:\nhttps://www.home-assistant.io/integrations/switcher_kis/#prerequisites",
```
```suggestion
        "description": "Found a Switcher device that requires a token\nEnter your username and token\nFor more information see https://www.home-assistant.io/integrations/switcher_kis/#prerequisites",
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,119 @@
+            if token_is_valid:
+                return await self._create_entry()
+            errors["base"] = "invalid_auth"
+        else:
+            user_input = {}
```
```suggestion
```

YogevBokobza replied:
It is done.

thecode commented on tests/components/switcher_kis/test_config_flow.py:
```diff
@@ -43,13 +50,90 @@ async def test_user_setup(
+        assert result2["type"] is FlowResultType.FORM
+        assert result2["step_id"] == "credentials"
+
+        with patch("aiohttp.ClientSession.post") as mock_post:
+            # Setup mock response for successful validation
+            mock_response = mock_post.return_value.__aenter__.return_value
+            mock_response.status = 200
+            mock_response.json.return_value = {"result": "true"}
```
Please patch the `validate_token` method to return true or false and not the `post` method
```suggestion
        with patch(
            "homeassistant.components.switcher_kis.config_flow.validate_token",
            return_value=True,
        ):
```

YogevBokobza replied:
It is done.

thecode commented on tests/components/switcher_kis/test_config_flow.py:
```diff
@@ -43,13 +50,90 @@ async def test_user_setup(
+        await hass.config_entries.async_remove(result3["result"].entry_id)
+        await hass.async_block_till_done()
+
+        # Now test with invalid credentials
```
Instead of reset the config entries, please split into two separate tests:
`test_user_setup_found_token_device_valid_token`
`test_user_setup_found_token_device_invalid_token`

thecode commented on tests/components/switcher_kis/test_config_flow.py:
```diff
@@ -43,13 +50,90 @@ async def test_user_setup(
+        assert result2["type"] is FlowResultType.FORM
+        assert result2["step_id"] == "credentials"
+
+        with patch("aiohttp.ClientSession.post") as mock_post:
+            # Setup mock response for failed validation
+            mock_response = mock_post.return_value.__aenter__.return_value
+            mock_response.status = 200
+            mock_response.json.return_value = {"result": "false"}
```
Patch `validate_token`

YogevBokobza replied:
It is done.

thecode commented on tests/components/switcher_kis/test_config_flow.py:
```diff
@@ -84,3 +168,82 @@ async def test_single_instance(hass: HomeAssistant) -> None:
+    )
+    entry.add_to_hass(hass)
+
+    with patch("aiohttp.ClientSession.post") as mock_post:
+        # Setup mock response for successful validation
+        mock_response = mock_post.return_value.__aenter__.return_value
+        mock_response.status = 200
+        mock_response.json.return_value = {"result": "true"}
```
Patch `validate_token`

thecode commented on tests/components/switcher_kis/test_cover.py:
```diff
@@ -182,6 +188,174 @@ async def test_cover_control_fail(hass: HomeAssistant, mock_bridge, mock_api) ->
+
+
+@pytest.mark.parametrize("mock_bridge", [[DEVICE2]], indirect=True)
+async def test_cover2(
```
This test is an exact copy of the `test_cover` with only change in the `entity_id`, please parameterize the `entity_id` in the existing `test_cover` test

thecode reviewed (changes requested):
Please address the comments, try to focus on the tests and avoid unnecessary changes not related to the tests.

YogevBokobza committed:
switcher fix based on requested changes

YogevBokobza commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -43,7 +43,12 @@ async def async_setup_entry(
+            async_add_entities([SwitcherCoverEntity(coordinator, 0)])
+        if (
+            coordinator.data.device_type.category
+            == DeviceCategory.SINGLE_SHUTTER_DUAL_LIGHT
```
It is done.

YogevBokobza commented on homeassistant/components/switcher_kis/config_flow.py:
```diff
@@ -2,9 +2,119 @@
+
+        for device_id, device in self.discovered_devices.items():
+            if device.token_needed:
+                _LOGGER.info("Device with ID %s requires a token", device_id)
```
It is done.

YogevBokobza committed:
switcher fix based on requested changes

YogevBokobza committed:
switcher fix based on requested changes

thecode commented on tests/components/switcher_kis/test_config_flow.py:
```diff
@@ -84,3 +174,78 @@ async def test_single_instance(hass: HomeAssistant) -> None:
+    )
+    entry.add_to_hass(hass)
+
+    with patch(
+        "homeassistant.components.switcher_kis.config_flow.validate_token",
+        return_value=True,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={
+                "source": config_entries.SOURCE_REAUTH,
+                "entry_id": entry.entry_id,
+            },
+            data=entry.data,
+        )
+
+        assert result["type"] is FlowResultType.FORM
+        assert result["step_id"] == "reauth_confirm"
+
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            user_input=user_input,
+        )
+
+        assert result["type"] is FlowResultType.ABORT
+        assert result["reason"] == "reauth_successful"
```
```suggestion
    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={
            "source": config_entries.SOURCE_REAUTH,
            "entry_id": entry.entry_id,
        },
        data=entry.data,
    )

    assert result["type"] is FlowResultType.FORM
    assert result["step_id"] == "reauth_confirm"

    with patch(
        "homeassistant.components.switcher_kis.config_flow.validate_token",
        return_value=True,
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            user_input=user_input,
        )

        assert result["type"] is FlowResultType.ABORT
        assert result["reason"] == "reauth_successful"
```

thecode reviewed (changes requested):
Minor adjustments to tests

YogevBokobza committed:
Update tests/components/switcher_kis/test_config_flow.py

Co-authored-by: Shay Levy <levyshay1@gmail.com>

YogevBokobza committed:
Update tests/components/switcher_kis/test_config_flow.py

Co-authored-by: Shay Levy <levyshay1@gmail.com>

YogevBokobza committed:
Update tests/components/switcher_kis/test_config_flow.py

Co-authored-by: Shay Levy <levyshay1@gmail.com>

thecode commented on tests/components/switcher_kis/test_config_flow.py:
```diff
@@ -84,3 +174,78 @@ async def test_single_instance(hass: HomeAssistant) -> None:
+            user_input=user_input,
+        )
+
+        assert result["type"] is FlowResultType.ABORT
+        assert result["reason"] == "reauth_successful"
```
```suggestion
    assert result["type"] is FlowResultType.ABORT
    assert result["reason"] == "reauth_successful"
```

thecode committed:
Update tests/components/switcher_kis/test_config_flow.py

thecode reviewed (approved):
Thanks @YogevBokobza 👍 

Feel free to make a PR to add yourself as a code-owner to get notified on issues and PRs.

YogevBokobza commented:
> Thanks @YogevBokobza 👍
> 
> Feel free to make a PR to add yourself as a code-owner to get notified on issues and PRs.

Thanks :)
How to make that PR? what file need to be change?
Who can merge this PR?
